File: CS/SimpleWebReportCatalog/CustomReportStorageWebExtension.cs
#region #CustomReportStorage
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using DevExpress.XtraReports.UI;
using SimpleWebReportCatalog.CatalogDataTableAdapters;
// ...

namespace SimpleWebReportCatalog {
    public class CustomReportStorageWebExtension : DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension {
        private CatalogData catalogDataSet;
        private DataTable reportsTable;
        private ReportsTableAdapter reportsTableAdapter;
        
        
        public CustomReportStorageWebExtension() {
            catalogDataSet = new CatalogData();
            reportsTableAdapter = new ReportsTableAdapter();
            reportsTableAdapter.Fill(catalogDataSet.Reports);
            reportsTable = catalogDataSet.Tables["Reports"];
        }


        public override bool CanSetData(string url) {
            // Check if the URL is available in the report storage.
            return reportsTable.Rows.Find(int.Parse(url)) != null;
        }


        public override byte[] GetData(string url) {
            // Get the report data from the storage.
            DataRow row = reportsTable.Rows.Find(int.Parse(url));
            if(row == null) return null; 

            byte[] reportData = (Byte[])row["LayoutData"];
            return reportData;
        }


        public override Dictionary<string, string> GetUrls() {
          // Get URLs and display names for all reports available in the storage.
          return reportsTable.AsEnumerable()
                .ToDictionary<DataRow, string, string>(dataRow => ((Int32)dataRow["ReportID"]).ToString(),
                                                       dataRow => (string)dataRow["DisplayName"]);         
        }


        public override bool IsValidUrl(string url) {
            // Check if the specified URL is valid for the current report storage.
            // In this example, a URL should be a string containing a numeric value that is used as a data row primary key.
            int n;
            return int.TryParse(url, out n);
        }


        public override void SetData(XtraReport report, string url) {
            // Write a report to the storage under the specified URL.
            DataRow row = reportsTable.Rows.Find(int.Parse(url));

            if(row != null) {
                using(MemoryStream ms = new MemoryStream()) {
                    report.SaveLayoutToXml(ms);
                    row["LayoutData"] = ms.GetBuffer();                
                }
                reportsTableAdapter.Update(catalogDataSet);
                catalogDataSet.AcceptChanges();
            }
        }


        public override string SetNewData(XtraReport report, string defaultUrl) {
            // Save a report to the storage under a new URL. 
            // The defaultUrl parameter contains the report display name specified by a user.
            DataRow row = reportsTable.NewRow();

            row["DisplayName"] = defaultUrl;
            using(MemoryStream ms = new MemoryStream()) {
                report.SaveLayoutToXml(ms);
                row["LayoutData"] = ms.GetBuffer();
            }

            reportsTable.Rows.Add(row);
            reportsTableAdapter.Update(catalogDataSet);
            catalogDataSet.AcceptChanges();

            // Refill the dataset to obtain the actual value of the new row's autoincrement key field.
            reportsTableAdapter.Fill(catalogDataSet.Reports);
            return catalogDataSet.Reports.FirstOrDefault(x => x.DisplayName == defaultUrl).ReportID.ToString();
        }
    }
}
#endregion #CustomReportStorage

File: VB/SimpleWebReportCatalog/CustomReportStorageWebExtension.vb
'#Region "#CustomReportStorage"
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.IO
Imports System.Linq
Imports DevExpress.XtraReports.UI
Imports SimpleWebReportCatalog.CatalogDataTableAdapters

' ...
Namespace SimpleWebReportCatalog

    Public Class CustomReportStorageWebExtension
        Inherits DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension

        Private catalogDataSet As CatalogData

        Private reportsTable As DataTable

        Private reportsTableAdapter As ReportsTableAdapter

        Public Sub New()
            catalogDataSet = New CatalogData()
            reportsTableAdapter = New ReportsTableAdapter()
            reportsTableAdapter.Fill(catalogDataSet.Reports)
            reportsTable = catalogDataSet.Tables("Reports")
        End Sub

        Public Overrides Function CanSetData(ByVal url As String) As Boolean
            ' Check if the URL is available in the report storage.
            Return reportsTable.Rows.Find(Integer.Parse(url)) IsNot Nothing
        End Function

        Public Overrides Function GetData(ByVal url As String) As Byte()
            ' Get the report data from the storage.
            Dim row As DataRow = reportsTable.Rows.Find(Integer.Parse(url))
            If row Is Nothing Then Return Nothing
            Dim reportData As Byte() = CType(row("LayoutData"), Byte())
            Return reportData
        End Function

        Public Overrides Function GetUrls() As Dictionary(Of String, String)
            ' Get URLs and display names for all reports available in the storage.
            Return reportsTable.AsEnumerable().ToDictionary(Of DataRow, String, String)(Function(dataRow)(CInt(dataRow(CStr("ReportID")))).ToString(), Function(dataRow) CStr(dataRow("DisplayName")))
        End Function

        Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
            ' Check if the specified URL is valid for the current report storage.
            ' In this example, a URL should be a string containing a numeric value that is used as a data row primary key.
            Dim n As Integer
            Return Integer.TryParse(url, n)
        End Function

        Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
            ' Write a report to the storage under the specified URL.
            Dim row As DataRow = reportsTable.Rows.Find(Integer.Parse(url))
            If row IsNot Nothing Then
                Using ms As MemoryStream = New MemoryStream()
                    report.SaveLayoutToXml(ms)
                    row("LayoutData") = ms.GetBuffer()
                End Using

                reportsTableAdapter.Update(catalogDataSet)
                catalogDataSet.AcceptChanges()
            End If
        End Sub

        Public Overrides Function SetNewData(ByVal report As XtraReport, ByVal defaultUrl As String) As String
            ' Save a report to the storage under a new URL. 
            ' The defaultUrl parameter contains the report display name specified by a user.
            Dim row As DataRow = reportsTable.NewRow()
            row("DisplayName") = defaultUrl
            Using ms As MemoryStream = New MemoryStream()
                report.SaveLayoutToXml(ms)
                row("LayoutData") = ms.GetBuffer()
            End Using

            reportsTable.Rows.Add(row)
            reportsTableAdapter.Update(catalogDataSet)
            catalogDataSet.AcceptChanges()
            ' Refill the dataset to obtain the actual value of the new row's autoincrement key field.
            reportsTableAdapter.Fill(catalogDataSet.Reports)
            Return Enumerable.FirstOrDefault(Of CatalogData.ReportsRow)(catalogDataSet.Reports, CType(Function(x) CBool(Equals(x.DisplayName, defaultUrl)), Func(Of CatalogData.ReportsRow, Boolean))).ReportID.ToString()
        End Function
    End Class
End Namespace
'#End Region  ' #CustomReportStorage
